### PR TITLE
[SPARK-44718][SQL] Match ColumnVector memory-mode config default to OffHeapMemoryMode config value

### PR DESCRIPTION
**Pull request by majdyz**
### What changes were proposed in this pull request?

Set the column vector default memory mode to depend on the off-heap memory mode flag. This is to prevent a user from using Vectorized-Reader with an on-heap column-vector by default when the off-heap memory mode is enabled on the cluster.

### Why are the changes needed?
Avoid the unintentional usage of on-heap memory in vectorized-reader when off-heap memory mode is enabled by the user.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Manual & existing tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -502,10 +502,14 @@ object SQLConf {
   val COLUMN_VECTOR_OFFHEAP_ENABLED =
     buildConf("spark.sql.columnVector.offheap.enabled")
       .internal()
-      .doc("When true, use OffHeapColumnVector in ColumnarBatch.")
+      .doc("When true, use OffHeapColumnVector in ColumnarBatch. " +
+        s"Defaults to $MEMORY_OFFHEAP_ENABLED && $MEMORY_OFFHEAP_SIZE > 0.")
       .version("2.3.0")
       .booleanConf
-      .createWithDefault(false)
+      .createWithDefaultFunction(() => {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:509]: I think you can use `fallbackConf`

Comment (majdyz) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Updated. But now the off-heap size check is removed, I hope it's fine.

**Review by HyukjinKwon — APPROVED** [review]
LGTM cc @cloud-fan @Ngone51 @jiangxb1987

**Comment by majdyz** [comment]
The failing CIs seems to be unrelated test failures.

File: sql/core/src/test/scala/org/apache/spark/sql/ConfigColumnVectorModeDefaultSuite.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+
+class ConfigColumnVectorModeDefaultSuite extends SharedSparkSession {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ConfigColumnVectorModeDefaultSuite.scala:24]: it's probably too much to add a new test suite for it. We have tests for the config framework and `.fallbackConf` is already tested by `ConfigEntrySuite`. Shall we just remove it?

Comment (majdyz) [reply sql/core/src/test/scala/org/apache/spark/sql/ConfigColumnVectorModeDefaultSuite.scala]: Removed :)

**Review by cloud-fan — APPROVED** [review]

**Pull request closed without merging**